lib/rake/test.rake

```diff
@@ -72,17 +72,85 @@ namespace :test do
   task :dashboard_ci do
     Dir.chdir(dashboard_dir) do
       ChatClient.wrap('dashboard ruby unit tests') do
-        # Unit tests mess with the database so stop the service before running them
-        RakeUtils.stop_service CDO.dashboard_unicorn_name
-        RakeUtils.rake 'db:test:prepare'
         ENV['DISABLE_SPRING'] = '1'
         ENV['UNIT_TEST'] = '1'
         ENV['CODECOV_FLAGS'] = 'dashboard'
-        TestRunUtils.run_dashboard_tests
+        ENV['PARALLEL_TEST_FIRST_IS_1'] = '1'
+        # Parallel tests don't seem to run more quickly over 16 processes.
+        ENV['PARALLEL_TEST_PROCESSORS'] = '16' if RakeUtils.nproc > 16
+
+        # Hash of all seed-data content: All fixture files plus schema.rb.
+        fixture_path = "#{dashboard_dir}/test/fixtures/"
+        fixture_hash = Digest::MD5.hexdigest(
+          Dir["#{fixture_path}/{**,*}/*.yml"].
+            push(dashboard_dir('db/schema.rb')).
+            select(&File.method(:file?)).
+            sort.
+            map(&Digest::MD5.method(:file)).
+            join
+        )
+        CDO.log.info "Fixture hash: #{fixture_hash}"
+
+        # Try to fetch seed data from S3
+        bucket_name = 'cdo-build-package'
+        s3_key = "test_db/#{fixture_hash}.gz"
+        s3_client = Aws::S3::Client.new
+        require 'zlib'
+        require 'stringio'
+
+        seed_data = begin
+          response = s3_client.get_object(bucket: bucket_name, key: s3_key)
+          Zlib::GzipReader.new(response.body).read
+        rescue Aws::Errors::MissingCredentialsError, Aws::S3::Errors::ServiceError
+          CDO.log.info "Seed data not found on S3 at #{s3_key}"
+          nil
+        end
+
+        seed_file = Tempfile.new(['db_seed', '.sql'])
+        auto_inc = 's/ AUTO_INCREMENT=[0-9]*\b//'
+        writer = URI.parse(CDO.dashboard_db_writer || 'mysql://root@localhost')
+
+        if seed_data
+          File.write(seed_file, seed_data)
+        else
+          # Generate new DB contents
+          ENV['TEST_ENV_NUMBER'] = '1'
+          RakeUtils.rake_stream_output 'db:create db:test:prepare'
+          ENV.delete 'TEST_ENV_NUMBER'
```

Comment:
This is an interesting bit; it looks like [parallel_tests doesn't normally use `TEST_ENV_NUMBER` '1' for any process](https://github.com/grosser/parallel_tests#add-to-configdatabaseyml).  What's going on here?